src/details/ArborX_MinimumSpanningTree.hpp

```diff
@@ -504,6 +509,150 @@ void finalizeEdges(ExecutionSpace const &space, BVH const &bvh,
+{
+  KokkosExt::ScopedProfileRegion guard("ArborX::MST::update_sided_parents");
+
+  // Same as dendrogram alpha's standalone "updateSidedParents"
+  Kokkos::parallel_for(
+      "ArborX::MST::update_sided_parents",
+      Kokkos::RangePolicy<ExecutionSpace>(space, edges_start, edges_end),
+      KOKKOS_LAMBDA(int e) {
+        auto const &edge = edges(e);
+
+        // As the edge is within the same alpha vertex, labels of its vertices
+        // are the same, so can take either
+        int component = labels(edge.source);
+
+        int const vertex_offset = edges_mapping.extent_int(0);
+        int const alpha_edge_index = edges_mapping(component - vertex_offset);
+
+        auto const &alpha_edge = edges(alpha_edge_index);
+
+        if (edge < alpha_edge)
+        {
+          bool is_left_side = (labels(alpha_edge.source) == component);
+          sided_parents(e) =
+              2 * alpha_edge_index + static_cast<int>(is_left_side);
+        }
+        else
+        {
+          sided_parents(e) = FOLLOW_CHAIN_VALUE - alpha_edge_index;
+        }
+      });
+}
+
+template <class ExecutionSpace, class Labels, class OutEdges,
+          class EdgesMapping, class BVH, class Parents>
+void assignVertexParents(ExecutionSpace const &space, Labels const &labels,
+                         OutEdges const &out_edges,
+                         EdgesMapping const &edges_mapping, BVH const &bvh,
+                         Parents parents)
+{
+  auto const n = edges_mapping.extent_int(0) + 1;
+  int const vertices_offset = n - 1;
+
+  Kokkos::parallel_for(
+      "ArborX::MST::compute_vertex_parents",
+      Kokkos::RangePolicy<ExecutionSpace>(space, 0, n), KOKKOS_LAMBDA(int e) {
+        auto const &edge = out_edges(e);
+
+        int i = labels(edge.source());
+        parents(HappyTreeFriends::getLeafPermutationIndex(bvh, i) +
+                vertices_offset) = edges_mapping(i - vertices_offset);
+      });
+}
+
+template <typename ExecutionSpace, typename Edges, typename SidedParents,
+          typename Parents>
+void computeParents(ExecutionSpace const &space, Edges const &edges,
+                    SidedParents const &sided_parents, Parents &parents)
+{
+  KokkosExt::ScopedProfileRegion guard("ArborX::MST::compute_edge_parents");
+
+  using MemorySpace = typename SidedParents::memory_space;
+
+  auto num_edges = edges.size();
+
+  // Encode both sided parent and edge into long long
+  // This way, once we sort based on this value, edges with the same sided
+  // parent will already be sorted in increasing order.
+  // The main reason for using long long values is the performance when
+  // compared with sorting pairs. The second reason is that Kokkos's BinSort
+  // does not support custom comparison operators.
+  Kokkos::View<long long *, MemorySpace> keys(
+      Kokkos::view_alloc(space, Kokkos::WithoutInitializing,
+                         "ArborX::MST::keys"),
+      num_edges);
+
+  constexpr int shift = 32;
```

Comment:
This needs to be
```suggestion
  constexpr int shift = sizeof(int) * CHAR_BIT;
```
right?

Reply:
I guess. I don't really find it more intuitive. If we decide to change this, we have quite a few places where we may want to do it.

Reply:
It just looked inconsistent checking that `float` and `int` have the same size but not testing for the size of `int` itself here. If we do tricks exploiting the size of certain types, I would prefer using `int32_t` and `int64_t` instead in relevant places (or apply the suggestion).